components/proxy/src/main/java/com/hotels/styx/admin/handlers/ServiceProviderHandler.java

```diff
@@ -0,0 +1,127 @@
+import static com.hotels.styx.api.HttpResponseStatus.NOT_FOUND;
+import static com.hotels.styx.api.HttpResponseStatus.NO_CONTENT;
+import static com.hotels.styx.api.HttpResponseStatus.OK;
+import static com.hotels.styx.infrastructure.configuration.json.ObjectMappers.addStyxMixins;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.stream.Collectors.joining;
+import static org.slf4j.LoggerFactory.getLogger;
+
+/**
+ * Admin interface access to service provider information, esp. Health Check monitors.
+ */
+public class ServiceProviderHandler implements WebServiceHandler {
+
+    private static final Logger LOGGER = getLogger(ServiceProviderHandler.class);
+
+    private static final ObjectMapper YAML_MAPPER = addStyxMixins(new ObjectMapper(new YAMLFactory()))
+            .addMixIn(StyxObjectDefinition.class, ServiceProviderDefMixin.class)
+            .configure(FAIL_ON_UNKNOWN_PROPERTIES, false)
+            .configure(AUTO_CLOSE_SOURCE, true);
+
+    private final UrlPatternRouter urlRouter;
+
+    public ServiceProviderHandler(StyxObjectStore<ProviderObjectRecord> providerDatabase) {
+        urlRouter = new UrlPatternRouter.Builder()
+                .get("/admin/service/providers", (request, context) -> {
+                    String output = providerDatabase.entrySet()
+                            .stream()
+                            .map(entry -> serialise(entry.getKey(), entry.getValue()))
+                            .collect(joining("\n"));
+
+                    return Eventual.of(response(output.length() == 0 ? NO_CONTENT : OK)
+                            .body(output, UTF_8)
+                            .build());
+
+                })
+                .get("/admin/service/provider/:providerName", (request, context) -> {
+                    String name = placeholders(context).get("providerName");
+
+                    try {
+                        String object = providerDatabase.get(name)
+                                .map(record -> serialise(name, record))
+                                .orElseThrow(ResourceNotFoundException::new);
+
+                        return Eventual.of(response(OK).body(object, UTF_8).build());
+                    } catch (ResourceNotFoundException e) {
+                        return Eventual.of(response(NOT_FOUND).build());
+                    }
+                })
+                .build();
+    }
+
+    @Override
+    public Eventual<HttpResponse> handle(HttpRequest request, HttpInterceptor.Context context) {
+        return urlRouter.handle(request, context);
+    }
+
+    public static ObjectMapper yamlMapper() {
+        return YAML_MAPPER.copy();
+    }
+
+    private static String serialise(String name, ProviderObjectRecord record) {
+        JsonNode node = YAML_MAPPER
+                .valueToTree(new StyxObjectDefinition(name, record.getType(), ImmutableList.copyOf(record.getTags()), record.getConfig()));
```

Comment:
I don't know if the length of this line is going to be a problem anywhere...

Reply:
There's a bit much going on anyway. Will rearrange.